core/src/test/scala/org/broadinstitute/dsde/rawls/mock/MockSamDAO.scala

```diff
@@ -0,0 +1,155 @@
+          SamBillingProjectPolicyNames.owner,
+          SamBillingProjectPolicyNames.workspaceCreator).map(policyName => SamPolicyWithNameAndEmail(policyName, SamPolicy(Set.empty, Set.empty, Set.empty), WorkbenchEmail(policyName.value + "@example.com")))
+
+      case _ => Set.empty
+    } )
+
+  override def listUserPoliciesForResource(resourceTypeName: SamResourceTypeName, resourceId: String, userInfo: UserInfo): Future[Set[SamPolicyWithName]] = ???
+
+  override def listUserRolesForResource(resourceTypeName: SamResourceTypeName, resourceId: String, userInfo: UserInfo): Future[Set[String]] = ???
+
+  override def getPolicySyncStatus(resourceTypeName: SamResourceTypeName, resourceId: String, policyName: SamResourcePolicyName, userInfo: UserInfo): Future[SamPolicySyncStatus] = Future.successful(SamPolicySyncStatus("", WorkbenchEmail("foo@bar.com")))
+
+  override def getResourceAuthDomain(resourceTypeName: SamResourceTypeName, resourceId: String, userInfo: UserInfo): Future[Seq[String]] = Future.successful(Seq.empty)
+
+  override def requestAccessToManagedGroup(groupName: WorkbenchGroupName, userInfo: UserInfo): Future[Unit] = ???
+
+  override def getPetServiceAccountKeyForUser(googleProject: String, userEmail: RawlsUserEmail): Future[String] = Future.successful("""{"client_email": "pet-110347448408766049948@broad-dsde-dev.iam.gserviceaccount.com"}""")
+
+  override def getDefaultPetServiceAccountKeyForUser(userInfo: UserInfo): Future[String] = Future.successful("""{"client_email": "pet-110347448408766049948@broad-dsde-dev.iam.gserviceaccount.com"}""")
+
+  override def getStatus(): Future[SubsystemStatus] = Future.successful(SubsystemStatus(true, None))
+
+  override def listAllResourceMemberIds(resourceTypeName: SamResourceTypeName, resourceId: String, userInfo: UserInfo): Future[Set[UserIdInfo]] = ???
+
+  override def getAccessInstructions(groupName: WorkbenchGroupName, userInfo: UserInfo): Future[Option[String]] = ???
+}
+
+class CustomizableMockSamDAO(dataSource: SlickDataSource)(implicit executionContext: ExecutionContext) extends MockSamDAO(dataSource) {
+  val userEmails = new TrieMap[String, Option[String]]()
+  val invitedUsers = new TrieMap[String, String]()
+  val policies = new TrieMap[(SamResourceTypeName, String), TrieMap[SamResourcePolicyName, SamPolicyWithNameAndEmail]]()
+
+  val callsToAddToPolicy = new ConcurrentLinkedDeque[(SamResourceTypeName, String, SamResourcePolicyName, String)]()
+  val callsToRemoveFromPolicy = new ConcurrentLinkedDeque[(SamResourceTypeName, String, SamResourcePolicyName, String)]()
+
+  override def registerUser(userInfo: UserInfo): Future[Option[UserStatus]] = {
+    userEmails.put(userInfo.userEmail.value, Option(userInfo.userSubjectId.value))
+    Future.successful(Option(UserStatus(RawlsUser(userInfo.userSubjectId, userInfo.userEmail), Map.empty)))
+  }
+
+  override def getUserIdInfo(userEmail: String, userInfo: UserInfo): Future[Either[Unit, Option[UserIdInfo]]] = {
```

Comment:
the return type seems a bit suspicious...`Future[Option[UserIdInfo]]` is not enough?

Reply:
ah I see...I would make it an ADT in this case.
something like:
```
sealed trait GetUserIdInfoResult extends Product with Serializable
final case object NotFound
final case object BelongToManagedGroup
final case class BelongToUser(userInfo: UserInfo)
```
And return `Future[GetUserIdInfoResult]`...it'll be easier if in the future we support more cases too